ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -0,0 +1,319 @@
+    x2: &[S],
+) -> Result<(), Error> {
+    // compute hashes
+    // compute hash for x2
+    let hash_x2 = compute_row_hash::<S, _, _>(keys, x2);
+    // compute hash for C
+    let hash_c = compute_row_hash::<S, _, _>(
+        keys,
+        share_b_and_c.iter().map(ReplicatedSecretSharing::right),
+    );
+
+    // setup channels
+    let h1_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashH2toH1)
+        .set_total_records(TotalRecords::specified(1)?);
+    let h3_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashH3toH2)
+        .set_total_records(TotalRecords::specified(1)?);
+    let channel_h1 = &h1_ctx.send_channel::<Hash>(ctx.role().peer(Direction::Left));
+    let channel_h3 = &h3_ctx.recv_channel::<Hash>(ctx.role().peer(Direction::Right));
+
+    // send and receive hash
+    let ((), hash_h3) = try_join(
+        channel_h1.send(RecordId::FIRST, hash_c),
+        channel_h3.receive(RecordId::FIRST),
+    )
+    .await?;
+
+    // check x2
+    if hash_x2 != hash_h3 {
+        return Err(Error::ShuffleValidationFailed(format!(
+            "X2 is inconsistent: hash of x2: {hash_x2:?}, hash of y2: {hash_h3:?}"
+        )));
+    }
+
+    Ok(())
+}
+
+/// This is the verification function run by `H3`.
+/// `H3` computes the hash for `y1`, `y2` and `c`
+/// and sends `y1`, `c` to `H1` and `y2` to `H2`.
+///
+/// ## Errors
+/// Propagates network errors.
+async fn h3_verify<C: Context, S: BooleanArray>(
+    ctx: C,
+    keys: &[StdArray<Gf32Bit, 1>],
+    share_c_and_a: &[AdditiveShare<S>],
+    y1: &[S],
+    y2: &[S],
+) -> Result<(), Error> {
+    // compute hashes
+    // compute hash for y1
+    let hash_y1 = compute_row_hash::<S, _, _>(keys, y1);
+    // compute hash for y2
+    let hash_y2 = compute_row_hash::<S, _, _>(keys, y2);
+    // compute hash for C
+    let hash_c = compute_row_hash::<S, _, _>(
+        keys,
+        share_c_and_a.iter().map(ReplicatedSecretSharing::left),
+    );
+
+    // setup channels
+    let h1_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashesH3toH1)
+        .set_total_records(TotalRecords::specified(2)?);
+    let h2_ctx = ctx
+        .narrow(&OPRFShuffleStep::HashH3toH2)
+        .set_total_records(TotalRecords::specified(1)?);
+    let channel_h1 = &h1_ctx.send_channel::<Hash>(ctx.role().peer(Direction::Right));
+    let channel_h2 = &h2_ctx.send_channel::<Hash>(ctx.role().peer(Direction::Left));
+
+    // send and receive hash
+    let _ = try_join3(
+        channel_h1.send(RecordId::FIRST, hash_y1),
+        channel_h1.send(RecordId::from(1usize), hash_c),
+        channel_h2.send(RecordId::FIRST, hash_y2),
+    )
+    .await?;
+
+    Ok(())
+}
+
+/// This function computes for each item in the iterator the inner product with `keys`.
+/// It concatenates all inner products and hashes them.
+///
+/// ## Panics
+/// Panics when conversion from `BooleanArray` to `Vec<Gf32Bit` fails.
+fn compute_row_hash<S, B, I>(keys: &[StdArray<Gf32Bit, 1>], row_iterator: I) -> Hash
```

Comment:
you probably want similar optimizations as for `compute_hash` to avoid turbofish syntax. Maybe in this case it could be easier because you really want `S` and not `&S` (you clone it inside), so you can ask for it